lib/ansible/modules/cloud/azure/azure_rm_disk.py

```diff
@@ -0,0 +1,352 @@
+        id:
+          description: Disk resource path.
+          type: str
+          example: "/subscriptions/XXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXX/resourceGroups/Testing/providers/Microsoft.Compute/disks/foobar"
+        name:
+          description: Disk name.
+          type: str
+          example: "foobar"
+        location:
+          description: Location of the disk
+          type: str
+          example: "eastus"
+        size_gb:
+          description: size of the disk
+          type: int
+          example: 64
+        sku:
+          description: sku of the disk
+          type: str
+          example: "Standard_LRS"
+        managed_by:
+          description: VM name which the disk attached
+          type: str
+          example: "testvm001"
+        provisioning_state:
+          description: Success or failure of the provisioning event.
+          type: str
+          example: "Succeeded"
+'''  # NOQA
+
+from ansible.module_utils.azure_rm_common import AzureRMModuleBase, CIDR_PATTERN, azure_id_to_dict
+
+try:
+    from msrestazure.azure_exceptions import CloudError
+    from azure.mgmt.compute.models import DataDisk, Disk, DiskSku, \
+                                          DiskCreateOption, DiskCreateOptionTypes, CreationData, ManagedDiskParameters
+except ImportError:
+    # This is handled in azure_rm_common
+    pass
+
+
+
+def disk_to_dict(disk):
+    result = dict(
+        id=disk.id,
+        name=disk.name,
+        provisioning_state=disk.provisioning_state,
+        location=disk.location,
+        managed_by=get_vm_name(disk.managed_by),
+        sku=disk.sku.name.value,
+        tags=disk.tags,
+        size_gb=disk.disk_size_gb,
+    )
+    return result
+
+def get_vm_name(url):
+    url = url or ''
+    return url.split('/')[-1]
+
+class AzureRMDisk(AzureRMModuleBase):
+
+    def __init__(self):
+
+        self.module_arg_spec = dict(
+            resource_group=dict(type='str', required=True),
+            name=dict(type='str', required=True),
+            state=dict(type='str', default='present', choices=['present', 'absent']),
+            location=dict(type='str'),
+            size_gb=dict(type='int', required=True),
+            sku=dict(type='str', default='Premium_LRS', choices=['Premium_LRS', 'Standard_LRS']),
+            managed_by=dict(type='str')
+        )
+
+        self.results = dict(
+            changed=False,
+            state=dict()
+        )
+
+        self.resource_group = None
+        self.name = None
+        self.state = None
+        self.location = None
+        self.size_gb = 0
+        self.sku = None
+        self.managed_by = None
+        self.tags = None
+
+        super(AzureRMDisk, self).__init__(self.module_arg_spec,
+                                            supports_check_mode=True)
+
+    def exec_module(self, **kwargs):
+
+        for key in self.module_arg_spec:
+            setattr(self, key, kwargs[key])
+
+        if self.state == 'absent' and self.managed_by:

```

Comment:
do we have to check it here?
I assume Azure will throw error when deleting a disk which is attached to VM. So we don't have to add extra logics to increase the complexity.

Reply:
agree, i think we should not try to fix anything in the module. we shoud focus only on properly presenting errors from the SDK